Custom criteria test: particle count

EtherealMC-Bit opened the pull request:
This is a simple test file that shows that the simulation will stop based on user defined criteria. In this case the simulation stops when the total particle count goes above 2.5 million particles.

peterscherpelz committed:
Initial, totally-not-complete, diode setup script

Commit:
Renamed Ar to inert_gas, raised not implemented errors, removed Cs-ionization functionality

Commit:
set up grids in init_base()

Commit:
Made not implemented errors more helpful

Commit:
Used PICMI electrostatic solver in init_solver()

Commit:
Boundry condition parameter fix and MAX_GRID_SIZE_FACTOR field

Commit:
Turned two fields into class input variables and added init_warpx() function

- Turned diag_steps and total_timesteps into class input variables called
  DIAG_STEPS and TOTAL_TIMESTEPS.

Commit:
commented out unused gentools.SetupCalcs() function

Commit:
added sim_control variable

Commit:
removed old imports

Commit:
Attempt at running capacitive_discharge using diode_setup

Commit:
Calcualte RES_LENGTH if NX, NY, NZ are passed in

Commit:
Species and field diagnostics now defined in diode_setup

- Particle species can be passed in as an input argument
- The list of field diagnostic data to be tracked can be defined as
  an input parameter.

Commit:
uncommented and started porting directory functions

Commit:
Passed dimension and rz to diode template, removed unused stuff

peterscherpelz committed:
Significantly refactor testing and diode setup

Commit:
fixed several issues

peterscherpelz committed:
Fix issues in moving from diode along x to along z

Also force PERIOD to be specified at present, and make RES_LENGTH always
reflect z direction.

peterscherpelz committed:
Make tests run in `temp_files`, not `test_files`.

peterscherpelz committed:
Remove old test no longer needed.

Commit:
Initialized the simulation inside mwxrun instead of diode_setup

peterscherpelz committed:
Remove passing simulation to init_run.

Also fixed test_capacitive_discharge.py for full compatibility with
pytest, and check that a string from run completion is present.

Commit:
Custom criteria test: particle count

EtherealMC-Bit committed:
Merge branch 'memaster2' into mkieburtz/test_custom_criteria

Commit:
Removed unused psutil import

Commit:
Merge branch 'mkieburtz/test_custom_criteria' of github.com:ModernElectron/WarpX into mkieburtz/test_custom_criteria

Commit:
removed minerva import from test_install_fieldsolver

roelof-groenewald reviewed (commented):
@MKieburtz How long does this test take to run on your machine? We should try to keep tests quick, so with that thought maybe decrease the total particle count by an order of magnitude or 2. The test will be just as effective with a smaller number of total particles but it might run significantly faster.

peterscherpelz commented on mewarpx/mewarpx/diags_store/diag_base.py:
```diff
@@ -3,7 +3,7 @@
 
 import numpy as np
-import psutil
+#import psutil
```
Add a space after `#` please.

peterscherpelz commented on mewarpx/tests/test_custom_criteria.py:
```diff
@@ -0,0 +1,108 @@
+    # # make sure out isn't empty
+    # assert out
+# this is how I'm running it
+test_capacitive_discharge_multigrid(None, "Run2D")
```
@MKieburtz I'm guessing this is your problem right now. `pytest` automatically runs functions that begin with `test`, and in general modules should *not* execute code like this when they're just imported. Sorry, I fixed it last minute in the main diode template and forgot to point it out to you, but see everything below line 84 https://github.com/ModernElectron/WarpX/blob/b2a85ee6fb43e4a58a4f616388f5e74f33c118ae/mewarpx/tests/test_capacitive_discharge.py#L84 - the `if __name__` block is what solves your problem here; it runs the test if you call it with `python` but not if you call it with `pytest`.

EtherealMC-Bit commented:
> How long does this test take to run on your machine?

@roelof-groenewald it takes about 4 seconds on penning. I'll go ahead and lower the number of particles.

peterscherpelz commented on mewarpx/tests/test_custom_criteria.py:
```diff
@@ -0,0 +1,108 @@
+
+    # Include a random run number to allow parallel runs to not collide. Using
+    # python randint prevents collisions due to numpy rseed below
+    #testing_util.initialize_testingdir(name)
```
You do want to be using this function

peterscherpelz commented on mewarpx/tests/test_custom_criteria.py:
```diff
@@ -0,0 +1,108 @@
+        NX=16,
+        NZ=128,
+        # This gives equal spacing in x & z
+        PERIOD=D_CA * 16 / 128.0,
```
See NX and NY predefined at https://github.com/ModernElectron/WarpX/blob/b2a85ee6fb43e4a58a4f616388f5e74f33c118ae/mewarpx/tests/test_capacitive_discharge.py#L45 please use that.

In general please do a diff against the `test_capacitive_discharge` in `memaster` to make sure you pick up all changes from there.

peterscherpelz commented on mewarpx/tests/test_custom_criteria.py:
```diff
@@ -0,0 +1,108 @@
+        DIAG_INTERVAL=DIAG_INTERVAL,
+        NUMBER_PARTICLES_PER_CELL=[16, 32],
+        FIELD_DIAG_DATA_LIST=['rho_electrons', 'rho_he_ions', 'phi'],
+        SIMULATION=sim
```
Another change in newer test

peterscherpelz commented on mewarpx/tests/test_custom_criteria.py:
```diff
@@ -0,0 +1,108 @@
+
+    # Run the main WARP loop
+    while run.control.check_criteria():
+        run.SIMULATION.step(1)
```
Another change in newer test

Commit:
Fixed several issues with test

- Lowered particle nums to make the simulation run quicker
- Tweaked check_particle_nums termination criteria so that the simulation will terminate after 1 step.
- Defined NX and NZ outside of diode setup initialization
- Commented out Run2D_RZ test

Commit:
Use only 1 OMP thread

Commit:
removed 3d test, renamed custom test function

peterscherpelz commented on mewarpx/tests/test_custom_criteria.py:
```diff
@@ -90,19 +89,22 @@ def test_capacitive_discharge_multigrid(capsys, name):
-test_capacitive_discharge_multigrid(None, "Run2D")
\ No newline at end of file
+if __name__ == '__main__':
+    test_capacitive_discharge_custom_criteria(None, "Run2D")
```
Newline here please.

peterscherpelz reviewed (approved):
@MKieburtz I'll mark this approved on my part, though please do add the end-of-file newline on the test. Thanks for all the work!

Commit:
added newline at end of file

KZhu-ME reviewed (approved):
Looks good!

EtherealMC-Bit committed:
Merge remote-tracking branch 'origin/memaster' into mkieburtz/test_custom_criteria

EtherealMC-Bit committed:
Exclude examples directory from pytest tests
